d2/src/test/java/com/linkedin/d2/balancer/dualread/UriPropertiesDualReadMonitorTest.java

```diff
@@ -0,0 +1,273 @@
+    monitor.reportData(CLUSTER_2, URI_PROPERTIES_URI_3_ANOTHER_DIFF_SPECIFIC_PROPERTIES_AND_4_DIFF_WEIGHT, false);
+    assertEquals(monitor.getTotalUris(), 4);
+    assertEquals(monitor.getMatchedUris(), 3);
+    verifyJmxMetricParams(fixture, CLUSTER_2,
+        createMatchRecord(URI_PROPERTIES_URI_3_ANOTHER_DIFF_SPECIFIC_PROPERTIES_AND_4_DIFF_WEIGHT,
+            URI_PROPERTIES_URI_3_DIFF_SPECIFIC_PROPERTIES_AND_4_DIFF_WEIGHT,
+            2, 1),
+        0.75);
+
+    // delete both lbs data for cluster 2
+    monitor.reportData(CLUSTER_2, null, true);
+    monitor.reportData(CLUSTER_2, null, false);
+    verifyJmxMetricParams(fixture, CLUSTER_2, null, 1.0);
+  }
+
+  @DataProvider
+  public Object[][] reportDataInMultiThreadsDataProvider() {
+    return new Object[][]{
+        // 2 updates
+        {
+            new ConcurrentLinkedDeque<>(Arrays.asList(
+                URI_PROPERTIES_1,
+                URI_PROPERTIES_URI_1_AND_2)),
+            new ConcurrentLinkedDeque<>(Arrays.asList(
+                URI_PROPERTIES_1,
+                URI_PROPERTIES_URI_1_AND_2))
+        },
+        // 3 updates
+        {
+            new ConcurrentLinkedDeque<>(Arrays.asList(
+                URI_PROPERTIES_1,
+                URI_PROPERTIES_URI_1_AND_2,
+                URI_PROPERTIES_1)),
+            new ConcurrentLinkedDeque<>(Arrays.asList(
+                URI_PROPERTIES_1,
+                URI_PROPERTIES_URI_1_AND_2,
+                URI_PROPERTIES_1))
+        },
+        // 5 updates
+        {
+            new ConcurrentLinkedDeque<>(Arrays.asList(
+                URI_PROPERTIES_1,
+                URI_PROPERTIES_URI_1_AND_2,
+                URI_PROPERTIES_2,
+                URI_PROPERTIES_URI_1_AND_2,
+                URI_PROPERTIES_1)),
+            new ConcurrentLinkedDeque<>(Arrays.asList(
+                URI_PROPERTIES_1,
+                URI_PROPERTIES_URI_1_AND_2,
+                URI_PROPERTIES_2,
+                URI_PROPERTIES_URI_1_AND_2,
+                URI_PROPERTIES_1))
+        }
+    };
+  }
+
+  @Test(dataProvider = "reportDataInMultiThreadsDataProvider", invocationCount = 20)
+  public void testReportDataInMultiThreads(Queue<UriProperties> oldLbProps, Queue<UriProperties> newLbProps) {
+    UriPropertiesDualReadMonitorTestFixture fixture = new UriPropertiesDualReadMonitorTestFixture();
+    UriPropertiesDualReadMonitor monitor = fixture.getMonitor();
+
+    ScheduledExecutorService executor = fixture.getExecutor();
+
+    CountDownLatch done = fixture.getDoneSignal(oldLbProps.size() + newLbProps.size());
+    // randomly report properties from old and new Lbs but properties in same lb are reported in order
+    executor.execute(() -> runNext(fixture, oldLbProps, false));
+    executor.execute(() -> runNext(fixture, newLbProps, true));
+
+    try {
+      boolean completed = done.await(1000, TimeUnit.MILLISECONDS);
+      if (completed) {
```

Comment:
Technically, if `!completed` the test should fail right? You're better off doing a bar `done.await()` and adding the timeout flag in the `@Test` annotation. That way you don't have to explicitly handle the failure case. Similarly, I would add `throws Exception` to the test signature and avoid the explicit `catch (InterruptedException e)`. It should make the test easier to read

Reply:
Actually it does time out rarely, so I added the “completed” check. I didn’t see anything wrong in the code though, do you?

Reply:
turns out it was because the test has multi-thread accessing some non-thread safe vars in the monitor. Made the getter synchronized.